Create OpenTelemetry collector docs for infrastructure monitoring (#6682)
newrelic/docs-website: src/content/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/collector/opentelemetry-collector-infra-hosts.mdx

The collector setup is part of the larger process of setting up OpenTelemetry with New Relic. Before following the collector steps below, make sure you've completed the [preceding setup steps](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-quick-start).

If you're looking for help with other collector use cases, see the [newrelic-opentelemetry-examples](https://github.com/newrelic/newrelic-opentelemetry-examples) repository.

## Step 1: Install the OpenTelemetry collector [#install-generic]

To do a basic installation for single hosts in the cloud or on-premises, see [OpenTelemetry's instructions](https://opentelemetry.io/docs/collector/getting-started/#linux-packaging) for up-to-date installation steps from the community. Instructions are available for the following:

* Linux: Debian systems
* Linux: Red Hat
* Windows
* Docker, Kubernetes, and other options

Your deployment experience might vary depending on which vendor-specific distributions you use. For example, installation via a package manager might be available for Linux hosts.

<Callout variant="important">

> To set up infrastructure monitoring, you need to install and configuring components that are included in the `collector-contrib` release. For example, the host receiver is required to collect basic host metrics such as CPU, memory, disk and network stats and is only available in the [OpenTelemetry collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib) release.

Reviewer: One collector-contrib reference in this paragraph has backticks, the other does not. I don't care which you use, but I do like consistency! :-)

Author: Barb, I think the instance without back-ticks is a link, so it may look unusual to have the back-ticks inside the link.

> To set up infrastructure monitoring, you need to install and configuring components that are included in the `collector-contrib` release. For example, the host receiver is required to collect basic host metrics such as CPU, memory, disk and network stats and is only available in the [OpenTelemetry collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib) release.

Reviewer 2: ```suggestion
To set up infrastructure monitoring, you need to install and configure components that are included in the `collector-contrib` release. For example, the host receiver is required to collect basic host metrics such as CPU, memory, disk and network stats and is only available in the [OpenTelemetry collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib) release.
```

</Callout>

## Step 2: Configure the OpenTelemetry collector for New Relic [#configure]

These examples are meant to serve as a starting point from which you can extend, customize, and validate configurations before using them in production.

### Host monitoring using the host receiver [#host-receiver]

> The `collector-contrib` release provides a `hostreceiver` that generates metrics about the system scraped from various sources. It is intended to be used when the collector is deployed as an agent.

Reviewer: Second sentence maybe change to more active voice, like "Use this option when you want to deploy the collector as an agent." or "We recommend that you use this option" etc

Author: @barbnewrelic, I went with this: Deploy the collector as an agent when you use a `hostreceiver`.

When using the host receiver as part of the collector configuration, New Relic automatically detects host metrics as part of a `Host` entity and will synthesize its golden metrics providing the same experience as with the New Relic infrastructure agent. The following are the configuration requirements to enable the `Host` entity experience in New Relic UI:

  * `host.id` attribute is present in host metrics.
  * `service.name` and `container.id` attributes are not present in host metrics.

> Learn more about available metrics and advanced configurations from the [OpenTelemetry documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/hostmetricsreceiver).

Reviewer: maybe say [OpenTelemetry documentation in GitHub] so readers know you don't mean OTEL docs on docs site

Author: @barbnewrelic, thanks! I made this change!

Adapt the `config.yaml` with these recommended parameters:

<Callout variant="important">

> CPU, load, memory and disk utilization metrics require otelcol-contri release `v47` or greater.

Reviewer: memory, and 
Oxford comma

Author: @barbnewrelic, thanks for the reminder!

</Callout>

<table>
  <thead>
    <tr>
      <th style={{ width: "290px" }}>
        Configuration
      </th>
      <th>
        Description
      </th>
    </tr>
  </thead>
    <tbody>
    <tr>
      <td>
        `receivers::hostmetrics`
      </td>
      <td>

>         Enable host metrics, 20 seconds interval is recommended (same default as the Infrastructure Agent), it should not be greater than 60 seconds to avoid issues with “host not responding” alerts. Process instrumentation is optional.

Reviewer: infrastructure agent l/c
First sentence seems like it really ought to be two

Author: @barbnewrelic, I divided these up into a bullet list!

      </td>
    </tr>
    <tr>
      <td>
        `processors::resourcedetection`
      </td>
      <td>
        Keep the following in mind:

          * `env`: Reads resource information from the `OTEL_RESOURCE_ATTRIBUTES` environment variable.
          * `system`: Adds `host.name` and `os.type`.

>           * For cloud environments, a specific resource detection processor needs to be configured so that metrics are decorated with `host.id` (required to identify the Host entity in New Relic One). Common cloud detectors are `gce` for GCP machines, `ec2` for AWS EC2 and `azure` for Azure VMs. Additional processors are available for orchestrated environments.

Reviewer: I'm not clear why Host needs to be initial cap?
Also:
AWS EC2, and azure
Oxford comma

Author: @barbnewrelic, two more helpful comments! Thanks!

>           * For on-premises systems (or unsupported cloud environments), a `host.id` attribute is required. The `resource` processor can be used to copy the `host.name` value (from `system`) as a new `host.id` attribute. Note this value should be unique across all instrumented hosts:

Reviewer: If you feel like revisiting passive voice throughout, this could be something like
Use the resource processor to copy...etc. Make sure this value is unique across all instrumented hosts.
I always worry when we use the word "should" does that mean, we don't require this, so you can do what you want at your peril? :-)

Author: @barbnewrelic, I fixed the two instance of passive voice in the table row. Thanks!

          ```
          resource:
              attributes:
                - key: host.id
                  from_attribute: host.name
                  action: upsert
          ```

      </td>
    </tr>
    <tr>
      <td>
        `processors::batch`
      </td>
      <td>
        The [batch processor](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/batchprocessor/README.md) accepts spans, metrics, or logs and places them into batches. Batching helps better compress the data and reduce the number of outgoing connections required to transmit the data. This processor supports both size and time based batching.
      </td>
    </tr>
    <tr>
      <td>
        `processors::memory_limited`
      </td>
      <td>

>         The [memory limiter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor/memorylimiterprocessor) processor is used to prevent out of memory situations on the collector. Given that the amount and type of data the collector processes is environment specific and resource utilization of the collector is also dependent on the configured processors, it is important to put checks in place regarding memory usage.

Reviewer: If this was my doc, I might reword last paragraph to something like this:
Putting checks in place is important because:
* The amount and type of data the collector processes is specific to its environment.
* The collector's resource utilization is dependent on the configured processors.

Author: @barbnewrelic, I used your solution. This is a nice streamlined approach!

      </td>
    </tr>
    <tr>
      <td>
        `processors::cumulative_delta`
      </td>
      <td>

>         The [cumulative delta](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/cumulativetodeltaprocessor) processor converts cumulative sum metrics to cumulative delta which is recommended in order to query system metric rates more easily in New Relic One.

Reviewer: maybe end this sentence after "delta." Then next sentence begins "This helps you query system metric rates" etc

Author: @barbnewrelic, good suggestion!

          system.memory.utilization:
            enabled: true
      disk:
      filesystem:
        metrics:
          system.filesystem.utilization:
            enabled: true
      network:
      paging:
        metrics:
          system.paging.utilization:
            enabled: true
      processes:
      process:

processors:
  memory_limiter:
    check_interval: 1s
    limit_mib: 1000
    spike_limit_mib: 200
  batch:
  cumulativetodelta:
    metrics:
      - system.network.io
      - system.disk.operations
      - system.network.dropped
      - system.network.packets
      - process.cpu.time
  resource:
    attributes:
      - key: host.id
        from_attribute: host.name
        action: upsert
  resourcedetection:
    detectors: [env, system]

exporters:
  otlp:
    endpoint: <var>OTLP_ENDPOINT_HERE</var>
    headers:
      api-key: <var>YOUR_KEY_HERE</var>

service:
  pipelines:
    metrics:
      receivers: [hostmetrics]
      processors: [batch, resourcedetection, resource, cumulativetodelta]
      exporters: [otlp]

  extensions: [health_check]
```

## Step 3: View your data [#using-ui]

You can view your collector data in a variety of places in the New Relic UI.

> ### Browsing host data in infrastructure UI [#using-ui]

Reviewer: anchor name on lines 200 and 204 are the same

Author: @barbnewrelic, good catch! I  just fixed this!

By using the recommended configuration for the host receiver, you can view data through the standard features in the [Infrastructure UI](/docs/infrastructure/infrastructure-ui-pages/infrastructure-ui-entities#access-new-ui) (New Host UI) experience.

### Query host metrics [#query-host-metrics]

> Once metrics are successfully ingested in New Relic, they are available in [data explorer](/docs/query-your-data/explore-query-data/browse-data/introduction-data-explorer) and [query builder](/docs/query-your-data/explore-query-data/query-builder/introduction-query-builder).

Reviewer: available in "the" data explorer?

Author: @barbnewrelic, I added "the."

The following NRQL queries show examples to help you explore the metrics you received:

  * Listing number of metric updates ingested by metric name

    ```
    SELECT count(*) FROM Metric WHERE metricName LIKE 'system.%' FACET metricName LIMIT max
    ```

  * Querying specific metrics faceted by host

    ```
    SELECT average(system.disk.operations) FROM Metric FACET host.name TIMESERIES
    ```

  * Listing dimensions available for a given metric

    ```
    SELECT keyset() FROM Metric WHERE metricName = 'system.disk.operations' 
    ```

Learn more about [querying the metric data type](/docs/data-apis/understand-data/metric-data/query-metric-data-type).

## What's next? [#next]

> After you set up your collector, check out our [best-practices](/docs/integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-concepts/) guide for tips to improve your use of OpenTelemetry and New Relic.

Reviewer: i don't think there needs to be a dash in "best-practices"

Author: @barbnewrelic, thanks! I removed it!

### Host monitoring using the host receiver [#host-receiver]

The `collector-contrib` release provides a `hostreceiver` that generates metrics about the system scraped from various sources. It is intended to be used when the collector is deployed as an agent.

When using the host receiver as part of the collector configuration, New Relic automatically detects host metrics as part of a `Host` entity and will synthesize its golden metrics providing the same experience as with the New Relic infrastructure agent. The following are the configuration requirements to enable the `Host` entity experience in New Relic UI:

  * `host.id` attribute is present in host metrics.
  * `service.name` and `container.id` attributes are not present in host metrics.

Learn more about available metrics and advanced configurations from the [OpenTelemetry documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/hostmetricsreceiver).

Adapt the `config.yaml` with these recommended parameters:

<Callout variant="important">

> CPU, load, memory and disk utilization metrics require otelcol-contri release `v47` or greater.

Reviewer 3: ```suggestion
CPU, load, memory and disk utilization metrics require otelcol-contrib release `v0.47.0` or greater.
```

> CPU, load, memory and disk utilization metrics require otelcol-contri release `v47` or greater.

Reviewer 3: ```suggestion
CPU, load, memory and disk utilization metrics require otelcol-contrib release `v0.47.0` or greater.
```

      </td>
    </tr>
    <tr>
      <td>
        `processors::memory_limited`
      </td>
      <td>
        The [memory limiter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor/memorylimiterprocessor) processor is used to prevent out of memory situations on the collector. Given that the amount and type of data the collector processes is environment specific and resource utilization of the collector is also dependent on the configured processors, it is important to put checks in place regarding memory usage.
      </td>
    </tr>
    <tr>
      <td>
        `processors::cumulative_delta`
      </td>
      <td>
        The [cumulative delta](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/cumulativetodeltaprocessor) processor converts cumulative sum metrics to cumulative delta which is recommended in order to query system metric rates more easily in New Relic One.
      </td>
    </tr>
    <tr>
      <td>
        `service::pipelines::metrics`
      </td>
      <td>
        Make sure `hostreceiver` and `resourcedetection` are included.
      </td>
    </tr>
    </tbody>
</table>

> Here is a sample configuration YAML file:

Reviewer 3: I'd add a note: Replace NEW_RELIC_OTLP_ENDPOINT and INGEST_LICENSE_KEY with... (same content from the basic setup)

> Here is a sample configuration YAML file:

Reviewer 3: I'd add a note: Replace NEW_RELIC_OTLP_ENDPOINT and INGEST_LICENSE_KEY with... (same content from the basic setup)

Author: @josemore, I added this right before the file example.

```
extensions:
  health_check:

receivers:
  hostmetrics:
    collection_interval: 20s
    scrapers:
      cpu:
        metrics:
          system.cpu.utilization:
            enabled: true
      load:
      memory:
        metrics:
          system.memory.utilization:
            enabled: true
      disk:
      filesystem:
        metrics:
          system.filesystem.utilization:
            enabled: true
      network:
      paging:
        metrics:
          system.paging.utilization:
            enabled: true
      processes:
      process:

processors:
  memory_limiter:
    check_interval: 1s
    limit_mib: 1000
    spike_limit_mib: 200
  batch:
  cumulativetodelta:
    metrics:
      - system.network.io
      - system.disk.operations
      - system.network.dropped
      - system.network.packets
      - process.cpu.time
  resource:
    attributes:
      - key: host.id
        from_attribute: host.name
        action: upsert
  resourcedetection:

>     detectors: [ env, system]

Reviewer 3: ```suggestion
    detectors: [env, system]
```

>     detectors: [ env, system]

Reviewer 3: ```suggestion
    detectors: [env, system]
```

Author: @josemore, I removed the space. GitHub seems to be having some problems today, so I wasn't able to accept your suggestion.

exporters: 
  otlp:

>     endpoint: otlp.nr-data.net:4317

Reviewer 3: ```suggestion
    endpoint: ***** NEW_RELIC_OTLP_ENDPOINT *****
```

>     endpoint: otlp.nr-data.net:4317

Reviewer 3: ```suggestion
    endpoint: ***** NEW_RELIC_OLTP_ENDPOINT *****
```

      exporters: [otlp]

  extensions: [health_check]
```

## Browsing host data in infrastructure UI [#using-ui]

Using the recommended configuration for the host receiver enables all the standard capabilities of the [Infrastructure UI](/docs/infrastructure/infrastructure-ui-pages/infrastructure-ui-entities#access-new-ui) (New Host UI) experience.

## Query host metrics [#query-host-metrics]

Once metrics are successfully ingested in New Relic, they are available in data explorer and query builder.

The following NRQL queries show examples to help you explore the metrics you received:

  * Listing number of metric updates ingested by metric name

    ```
    SELECT count(*) FROM Metric WHERE metricName LIKE 'system.%' FACET metricName LIMIT max
    ```

  * Querying specific metrics faceted by host

    ```
    SELECT average(system.disk.operations) FROM Metric FACET host.name TIMESERIES
    ```

  * Listing dimensions available for a given metric

    ```
    SELECT keyset() FROM Metric WHERE metricName = 'system.disk.operations' 
    ```

Learn more about [querying the metric data type](/docs/data-apis/understand-data/metric-data/query-metric-data-type).

> ## Additional settings [#additional-settings]

Reviewer 3: Starting from here till the end, I'd move it to a different doc focused on general settings for the collector & New Relic

Author: @josemore, I moved this to its own topic.

> ## Additional settings [#additional-settings]

Reviewer 3: Starting from here till the end, I'd move it to a different doc focused on general settings for the collector & New Relic

Author: @josemore, I moved this into its own topic.
